test/opt/pass_test.cpp

```diff
@@ -0,0 +1,281 @@
+       %Dead = OpFunction %float None %8
+         %19 = OpLabel
+               OpReturnValue %float_1
+               OpFunctionEnd
+         %12 = OpFunction %float None %9
+         %20 = OpLabel
+         %21 = OpFunctionCall %float %13
+               OpReturnValue %21
+               OpFunctionEnd
+         %13 = OpFunction %float None %8
+         %22 = OpLabel
+               OpReturnValue %float_1
+               OpFunctionEnd
+)";
+  // clang-format on
+
+  std::unique_ptr<ir::Module> module =
+      BuildModule(SPV_ENV_UNIVERSAL_1_1, nullptr, text,
+                  SPV_TEXT_TO_BINARY_OPTION_PRESERVE_NUMERIC_IDS);
+  EXPECT_NE(nullptr, module) << "Assembling failed for shader:\n"
+                             << text << std::endl;
+  DummyPass testPass;
+  std::vector<uint32_t> processed;
+  opt::Pass::ProcessFunction mark_visited = [&processed](ir::Function* fp) {
+    processed.push_back(fp->result_id());
+    return false;
+  };
+  testPass.ProcessReachableCallTree(mark_visited, module.get());
+  EXPECT_THAT(processed, UnorderedElementsAre(10, 11, 12, 13));
+}
+
+TEST_F(PassClassTest, BasicVisitOnlyOnce) {
+  // Make sure we visit %11 only once, even if it is called from two different
+  // functions.
+  const std::string text = R"(
+               OpCapability Shader
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %10 "main" %gl_FragColor
+               OpExecutionMode %10 OriginUpperLeft
+               OpSource GLSL 150
+               OpName %10 "main"
+               OpName %Dead "Dead"
+               OpName %11 "Constant"
+               OpName %gl_FragColor "gl_FragColor"
+               OpName %12 "ExportedFunc"
+               OpDecorate %gl_FragColor Location 0
+               OpDecorate %12 LinkageAttributes "ExportedFunc" Export
+       %void = OpTypeVoid
+          %6 = OpTypeFunction %void
+      %float = OpTypeFloat 32
+          %8 = OpTypeFunction %float
+    %v4float = OpTypeVector %float 4
+%_ptr_Output_v4float = OpTypePointer Output %v4float
+%gl_FragColor = OpVariable %_ptr_Output_v4float Output
+    %float_1 = OpConstant %float 1
+         %10 = OpFunction %void None %6
+         %14 = OpLabel
+         %15 = OpFunctionCall %float %11
+         %16 = OpFunctionCall %float %11
+         %17 = OpCompositeConstruct %v4float %15 %16 %float_1 %float_1
+               OpStore %gl_FragColor %17
+               OpReturn
+               OpFunctionEnd
+         %11 = OpFunction %float None %8
+         %18 = OpLabel
+               OpReturnValue %float_1
+               OpFunctionEnd
+       %Dead = OpFunction %float None %8
+         %19 = OpLabel
+               OpReturnValue %float_1
+               OpFunctionEnd
+         %12 = OpFunction %float None %9
+         %20 = OpLabel
+         %21 = OpFunctionCall %float %12
+               OpReturnValue %21
+               OpFunctionEnd
+)";
+  // clang-format on
+
+  std::unique_ptr<ir::Module> module =
```

Comment:
All the tests in this file have the same basic pattern.  You could encapsulate the following steps in a function: (a) assemble with preserve IDs, (b) run a dummy pass to collect visited ids, (c) return that list. 

That would be a nice cleanup, but optional.
Also, the tests do a bunch of work with various values.  It's usually best to eliminate all but absolutely necessary instructions.  The functionality in this test file is simple enough that it isn't a big problem, but you should get into the habit of pruning as much as possible.